File: messenger/app/src/test/java/com/example/messenger/crypto/DoubleRatchetTest.kt
package com.example.messenger.crypto

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Base64

class DoubleRatchetTest {
    @Test
    fun signedPrekeyVerifiesAndTamperingFails() {
        val bob = DoubleRatchet.createIdentity("bob", "phone")
        val bundle = bob.toPrekeyBundle()

        assertTrue(DoubleRatchet.verifySignedPrekey(bundle))

        val tampered = bundle.copy(signedPrekeyPublic = bundle.signedPrekeyPublic.copyOf().also { it[0] = (it[0].toInt() xor 1).toByte() })
        assertFalse(DoubleRatchet.verifySignedPrekey(tampered))
    }

    @Test
    fun encryptsAndDecryptsSequentialMessages() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val bob = DoubleRatchet.createIdentity("bob", "phone")

        val first = DoubleRatchet.encrypt("hello", alice, bob.toPrekeyBundle(), "alice|bob")
        val firstPlain = DoubleRatchet.decrypt(first, bob, alice.toIdentityKey(), "alice|bob")
        val second = DoubleRatchet.encrypt("again", alice, bob.toPrekeyBundle(), "alice|bob")
        val secondPlain = DoubleRatchet.decrypt(second, bob, alice.toIdentityKey(), "alice|bob")

        assertEquals("hello", firstPlain)
        assertEquals("again", secondPlain)
        assertEquals(DoubleRatchet.ALGORITHM, first.e2eeAlgorithm)
        assertEquals(first.messageNumber + 1, second.messageNumber)
    }

    @Test
    fun ratchetAdvancesAfterBothSidesSend() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val bob = DoubleRatchet.createIdentity("bob", "phone")

        val a1 = DoubleRatchet.encrypt("a1", alice, bob.toPrekeyBundle(), "alice|bob")
        assertEquals("a1", DoubleRatchet.decrypt(a1, bob, alice.toIdentityKey(), "alice|bob"))

        val b1 = DoubleRatchet.encrypt("b1", bob, alice.toPrekeyBundle(), "alice|bob")
        assertEquals("b1", DoubleRatchet.decrypt(b1, alice, bob.toIdentityKey(), "alice|bob"))

        val a2 = DoubleRatchet.encrypt("a2", alice, bob.toPrekeyBundle(), "alice|bob")
        assertEquals("a2", DoubleRatchet.decrypt(a2, bob, alice.toIdentityKey(), "alice|bob"))
        assertNotEquals(a1.ratchetPublicKey.toList(), a2.ratchetPublicKey.toList())
    }

    @Test
    fun exportImportPreservesSessionState() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val bob = DoubleRatchet.createIdentity("bob", "phone")

        val first = DoubleRatchet.encrypt("persist me", alice, bob.toPrekeyBundle(), "alice|bob")
        assertEquals("persist me", DoubleRatchet.decrypt(first, bob, alice.toIdentityKey(), "alice|bob"))

        val restored = DoubleRatchet.importIdentity(DoubleRatchet.exportIdentity(bob))
        val second = DoubleRatchet.encrypt("after restore", alice, bob.toPrekeyBundle(), "alice|bob")

        assertEquals("after restore", DoubleRatchet.decrypt(second, restored, alice.toIdentityKey(), "alice|bob"))
    }

    @Test
    fun exportImportPreservesSenderMessageKeysForOwnDecrypt() {
        val alice = DoubleRatchet.createIdentity("alice", "android")
        val bob = DoubleRatchet.createIdentity("bob", "web")

        val outbound = DoubleRatchet.encrypt("own copy", alice, bob.toPrekeyBundle(), "alice|bob")
        val restoredAlice = DoubleRatchet.importIdentity(DoubleRatchet.exportIdentity(alice))

        assertEquals("own copy", DoubleRatchet.decrypt(outbound, restoredAlice, bob.toIdentityKey(), "alice|bob"))
    }

    @Test
    fun decryptsOlderSkippedMessageAfterLaterMessage() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val bob = DoubleRatchet.createIdentity("bob", "android")

        val first = DoubleRatchet.encrypt("first", alice, bob.toPrekeyBundle(), "alice|bob")
        val second = DoubleRatchet.encrypt("second", alice, bob.toPrekeyBundle(), "alice|bob")

        assertEquals("second", DoubleRatchet.decrypt(second, bob, alice.toIdentityKey(), "alice|bob"))
        assertEquals("first", DoubleRatchet.decrypt(first, bob, alice.toIdentityKey(), "alice|bob"))
    }

    @Test
    fun decryptsIndependentInitialMessagesAfterBothSidesSendBeforeReceiving() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val bob = DoubleRatchet.createIdentity("bob", "android")

        val aliceFirst = DoubleRatchet.encrypt("alice first", alice, bob.toPrekeyBundle(), "alice|bob")
        val bobFirst = DoubleRatchet.encrypt("bob first", bob, alice.toPrekeyBundle(), "alice|bob")

        assertEquals("alice first", DoubleRatchet.decrypt(aliceFirst, bob, alice.toIdentityKey(), "alice|bob"))
        assertEquals("bob first", DoubleRatchet.decrypt(bobFirst, alice, bob.toIdentityKey(), "alice|bob"))
    }

    @Test
    fun rejectsInvalidSignedPrekeyBeforeEncrypting() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val bob = DoubleRatchet.createIdentity("bob", "phone")
        val bundle = bob.toPrekeyBundle()
        val invalid = bundle.copy(signedPrekeySignature = bundle.signedPrekeySignature.copyOf().also { it[0] = (it[0].toInt() xor 1).toByte() })

        try {
            DoubleRatchet.encrypt("nope", alice, invalid, "alice|bob")
            throw AssertionError("encrypt should have failed")
        } catch (expected: IllegalArgumentException) {
            assertTrue(expected.message!!.contains("invalid signed prekey signature", ignoreCase = true))
        }
    }

    @Test
    fun fixedVectorRoundTripKeepsBytesStableInsideJvm() {
        val alice = DoubleRatchet.createIdentity("alice", "web")
        val exported = DoubleRatchet.exportIdentity(alice)
        val restored = DoubleRatchet.importIdentity(exported)

        assertArrayEquals(alice.identityPublic, restored.identityPublic)
        assertArrayEquals(alice.signedPrekeySignature, restored.signedPrekeySignature)
    }

    @Test
    fun decryptsWebGeneratedMessageVector() {
        val bob = RatchetIdentity(
            username = "bob",
            deviceId = "android",
            keyId = "bob-identity-mpgmqel8-38g1fujk",
            identityPrivate = b64("ulRJy4aCCGre5tehpFqMqLGOGul0Qa0jKzKEkiLbsRc="),
            identityPublic = b64("rpL9jfX3gCYm4m8wCnRf6DjMCX5XM5rvT1mbNRMFrIw="),
            signedPrekeyId = "bob-spk-mpgmqel8-vadir4dx",
            signedPrekeyPrivate = b64("A2ON/Bj87DrwLbN78YdfBGhN6Ps9tBMWWZ9pa2p3WoI="),
            signedPrekeyPublic = b64("MYnbz6nN7fkQaU3wTEuyFIyyamVZXrO5eUGjxbBp4ic="),
            signedPrekeySignature = b64("4iDXby9ab9buSCQDo7LN2lU3vPeOIK5I5yoTEDWvoWtGVL14Qgp3Y2+ICEx1k5QsBqYC2wXwhZyohL+HJWixDQ=="),
        )
        val alice = RatchetIdentityKey(
            username = "alice",
            deviceId = "web",
            keyId = "alice-identity-mpgmqel6-j2pqqifa",
            publicKey = b64("75iCTjpdoNwMgwGUj/3lxOgpSBdpzY1DdZvge8LVCCg="),
        )
        val message = RatchetMessage(
            ciphertext = b64("Bj04vcxtq/bR07V7DuUtKbc3lAfvek2HX2uYptGq"),
            nonce = b64("gAx4abjbc+cSxcY7"),
            senderKeyId = alice.keyId,
            recipientSignedPrekeyId = bob.signedPrekeyId,
            recipientSignedPrekeyPublic = bob.signedPrekeyPublic,
            recipientOneTimePrekeyId = "",
            recipientOneTimePrekeyPublic = null,
            e2eeAlgorithm = DoubleRatchet.ALGORITHM,
            ratchetPublicKey = b64("k4Ntp0QarO37bcZcNfp4N3aknM1lbjPmciRjBngcURc="),
            previousChainLength = 0,
            messageNumber = 0,
        )

        assertEquals("web to android", DoubleRatchet.decrypt(message, bob, alice, "alice|bob"))
    }

    private fun b64(value: String): ByteArray = Base64.getDecoder().decode(value)
}

File: web/src/lib/double-ratchet.test.js
import { test } from "vitest";
import assert from "node:assert/strict";

import {
  DR_ALGORITHM,
  createRatchetIdentity,
  decryptRatchetMessage,
  encryptRatchetMessage,
  exportRatchetIdentity,
  importRatchetIdentity,
  verifySignedPrekey,
} from "./double-ratchet.js";

test("Ed25519 signed prekey verifies and tampering fails", async () => {
  const bob = await createRatchetIdentity("bob", "phone");
  const bundle = bob.toPrekeyBundle();

  assert.equal(await verifySignedPrekey(bundle), true);

  const tampered = {
    ...bundle,
    signedPrekey: {
      ...bundle.signedPrekey,
      publicKey: new Uint8Array(bundle.signedPrekey.publicKey),
    },
  };
  tampered.signedPrekey.publicKey[0] ^= 1;

  assert.equal(await verifySignedPrekey(tampered), false);
});

test("Double Ratchet decrypts first and sequential messages", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");

  const first = await encryptRatchetMessage("hello", alice, bob.toPrekeyBundle(), "alice|bob");
  const firstPlain = await decryptRatchetMessage(first, bob, alice.toIdentityKey(), "alice|bob");
  const second = await encryptRatchetMessage("again", alice, bob.toPrekeyBundle(), "alice|bob");
  const secondPlain = await decryptRatchetMessage(second, bob, alice.toIdentityKey(), "alice|bob");

  assert.equal(firstPlain, "hello");
  assert.equal(secondPlain, "again");
  assert.equal(first.e2eeAlgorithm, DR_ALGORITHM);
  assert.equal(second.messageNumber, first.messageNumber + 1);
});

test("Double Ratchet advances after both sides send", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");

  const a1 = await encryptRatchetMessage("a1", alice, bob.toPrekeyBundle(), "alice|bob");
  assert.equal(await decryptRatchetMessage(a1, bob, alice.toIdentityKey(), "alice|bob"), "a1");

  const b1 = await encryptRatchetMessage("b1", bob, alice.toPrekeyBundle(), "alice|bob");
  assert.equal(await decryptRatchetMessage(b1, alice, bob.toIdentityKey(), "alice|bob"), "b1");

  const a2 = await encryptRatchetMessage("a2", alice, bob.toPrekeyBundle(), "alice|bob");
  assert.equal(await decryptRatchetMessage(a2, bob, alice.toIdentityKey(), "alice|bob"), "a2");
  assert.notDeepEqual(Array.from(a1.ratchetPublicKey), Array.from(a2.ratchetPublicKey));
});

test("ratchet identity export/import preserves session state", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");

  const first = await encryptRatchetMessage("persist me", alice, bob.toPrekeyBundle(), "alice|bob");
  assert.equal(await decryptRatchetMessage(first, bob, alice.toIdentityKey(), "alice|bob"), "persist me");

  const restored = await importRatchetIdentity(await exportRatchetIdentity(bob));
  const second = await encryptRatchetMessage("after restore", alice, bob.toPrekeyBundle(), "alice|bob");

  assert.equal(await decryptRatchetMessage(second, restored, alice.toIdentityKey(), "alice|bob"), "after restore");
});

test("invalid signed prekey prevents encryption", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");
  const bundle = bob.toPrekeyBundle();
  bundle.signedPrekey.signature = new Uint8Array(bundle.signedPrekey.signature);
  bundle.signedPrekey.signature[0] ^= 1;

  await assert.rejects(
    () => encryptRatchetMessage("nope", alice, bundle, "alice|bob"),
    /invalid signed prekey signature/i,
  );
});

test("failed decrypt does not advance receiving ratchet session", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");

  const message = await encryptRatchetMessage("retryable", alice, bob.toPrekeyBundle(), "alice|bob");
  const tampered = {
    ...message,
    ciphertext: new Uint8Array(message.ciphertext),
  };
  tampered.ciphertext[0] ^= 1;

  await assert.rejects(
    () => decryptRatchetMessage(tampered, bob, alice.toIdentityKey(), "alice|bob"),
    /operation-specific reason|decrypt|failed/i,
  );

  assert.equal(await decryptRatchetMessage(message, bob, alice.toIdentityKey(), "alice|bob"), "retryable");
});

test("Double Ratchet decrypts older skipped messages after a later message", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");

  const first = await encryptRatchetMessage("first", alice, bob.toPrekeyBundle(), "alice|bob");
  const second = await encryptRatchetMessage("second", alice, bob.toPrekeyBundle(), "alice|bob");

  assert.equal(await decryptRatchetMessage(second, bob, alice.toIdentityKey(), "alice|bob"), "second");
  assert.equal(await decryptRatchetMessage(first, bob, alice.toIdentityKey(), "alice|bob"), "first");
});

test("Double Ratchet recovers an older initial-chain message if skipped cache was not persisted", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "phone");

  const first = await encryptRatchetMessage("first", alice, bob.toPrekeyBundle(), "alice|bob");
  const second = await encryptRatchetMessage("second", alice, bob.toPrekeyBundle(), "alice|bob");

  assert.equal(await decryptRatchetMessage(second, bob, alice.toIdentityKey(), "alice|bob"), "second");
  bob.skippedMessageKeys = {};

  assert.equal(await decryptRatchetMessage(first, bob, alice.toIdentityKey(), "alice|bob"), "first");
});

test("Double Ratchet decrypts independent initial messages after both sides send before receiving", async () => {
  const alice = await createRatchetIdentity("alice", "web");
  const bob = await createRatchetIdentity("bob", "android");

  const aliceFirst = await encryptRatchetMessage("alice first", alice, bob.toPrekeyBundle(), "alice|bob");
  const bobFirst = await encryptRatchetMessage("bob first", bob, alice.toPrekeyBundle(), "alice|bob");

  assert.equal(await decryptRatchetMessage(aliceFirst, bob, alice.toIdentityKey(), "alice|bob"), "alice first");
  assert.equal(await decryptRatchetMessage(bobFirst, alice, bob.toIdentityKey(), "alice|bob"), "bob first");
});
